_episodes/10-defensive.md

```diff
@@ -47,225 +49,229 @@ As in real carpentry --- the kind done with lumber --- the time saved
 by measuring carefully before cutting a piece of wood
 is much greater than the time that measuring takes.
 
-## Assertions
+## Raising errors
 
-The first step toward getting the right answers from our programs
-is to assume that mistakes *will* happen
-and to guard against them.
-This is called [defensive programming]({{ page.root }}/reference.html#defensive-programming),
-and the most common way to do it is to add
-[assertions]({{ page.root }}/reference.html#assertion) to our code
-so that it checks itself as it runs.
-An assertion is simply a statement that something must be true at a certain point in a program.
-When Python sees one,
-it evaluates the assertion's condition.
-If it's true,
-Python does nothing,
-but if it's false,
-Python halts the program immediately
-and prints the error message if one is provided.
-For example,
-this piece of code halts as soon as the loop encounters a value that isn't positive:
+One of the most commonly used tools in the defensive programming toolkit is the `raise` keyword,
+which can be used to trigger an error (otherwise known as "raising an exception").
 
 ~~~
-numbers = [1.5, 2.3, 0.7, -0.001, 4.4]
-total = 0.0
-for num in numbers:
-    assert num > 0.0, 'Data should only contain positive values'
-    total += num
-print('total is:', total)
+problem = True
+if problem:
+    raise Exception('There has been a problem')
 ~~~
 {: .language-python}
 
 ~~~
----------------------------------------------------------------------------
-AssertionError                            Traceback (most recent call last)
-<ipython-input-19-33d87ea29ae4> in <module>()
-      2 total = 0.0
-      3 for num in numbers:
-----> 4     assert num > 0.0, 'Data should only contain positive values'
-      5     total += num
-      6 print('total is:', total)
+Exception                                 Traceback (most recent call last)
+<ipython-input-2-44b57a133722> in <module>
+      1 if problem:
+----> 2     raise Exception('There has been a problem')
+      3 
 
-AssertionError: Data should only contain positive values
+Exception: There has been a problem
 ~~~
 {: .error}
 
-Programs like the Firefox browser are full of assertions:
-10-20% of the code they contain
-are there to check that the other 80–90% are working correctly.
-Broadly speaking,
-assertions fall into three categories:
-
-*   A [precondition]({{ page.root }}/reference.html#precondition)
-    is something that must be true at the start of a function in order for it to work correctly.
+In this case we've chosen to raise a generic `Exception`,
+but we could pick any of the [builtin exception types](https://docs.python.org/3/library/exceptions.html#exception-hierarchy)
+or define our own [custom exception class](https://towardsdatascience.com/how-to-define-custom-exception-classes-in-python-bfa346629bca).
```

Comment:
I would emphasise here that understanding how to define custom exception classes requires knowledge of concepts that are beyond the scope of this lesson.

Reply:
Fixed :smile: